libsql-sqlite3/src/vectorIndex.c

```diff
@@ -0,0 +1,462 @@
+    value >>= 8;
+  }
+  return 0;
+}
+
+double vectorIdxParamsGetF64(const VectorIdxParams *pParams, char tag) {
+  u64 value = vectorIdxParamsGetU64(pParams, tag);
+  return *((double*)&value);
+}
+
+int vectorIdxParamsPutF64(VectorIdxParams *pParams, char tag, double value) {
+  return vectorIdxParamsPutU64(pParams, tag, *((u64*)&value));
+}
+
+/**************************************************************************
+** VectorIdxKey utilities
+****************************************************************************/
+
+int vectorIdxKeyGet(Table *pTable, VectorIdxKey *pKey, const char **pzErrMsg) {
+  int i;
+  Index *pPk;
+  // we actually need to change strategy here and use PK if it's available and fallback to ROWID only if there is no other choice
+  // will change this later as it must be done carefully in order to not brake behaviour of existing indices
+  if( !HasRowid(pTable) ){
+    pPk = sqlite3PrimaryKeyIndex(pTable);
+    if( pPk->nKeyCol > VECTOR_INDEX_MAX_KEY_COLUMNS ){
+      *pzErrMsg = "exceeded limit for composite columns in primary key index";
+      return -1;
+    }
+    pKey->nKeyColumns = pPk->nKeyCol;
+    for(i = 0; i < pPk->nKeyCol; i++){
+      pKey->aKeyAffinity[i] = pTable->aCol[pPk->aiColumn[i]].affinity;
+      pKey->azKeyCollation[i] = pPk->azColl[i];
+    }
+  } else{
+    pKey->nKeyColumns = 1;
+    pKey->aKeyAffinity[0] = SQLITE_AFF_INTEGER;
+    pKey->azKeyCollation[0] = "BINARY";
+  }
+  return 0;
+}
+
+int vectorIdxKeyDefsRender(const VectorIdxKey *pKey, const char *prefix, char *pBuf, int nBufSize) {
+  static const char * const azType[] = {
+    /* SQLITE_AFF_BLOB    */ " BLOB",
+    /* SQLITE_AFF_TEXT    */ " TEXT",
+    /* SQLITE_AFF_NUMERIC */ " NUMERIC",
+    /* SQLITE_AFF_INTEGER */ " INTEGER",
+    /* SQLITE_AFF_REAL    */ " REAL",
+    /* SQLITE_AFF_FLEXNUM */ " NUMERIC",
+  };
+  int i, size;
+  for(i = 0; i < pKey->nKeyColumns && nBufSize > 0; i++){
+    const char *collation = pKey->azKeyCollation[i];
+    if( sqlite3_strnicmp(collation, "BINARY", 6) == 0 ){
+      collation = "";
+    }
+    if( i == 0 ){
+      size = snprintf(pBuf, nBufSize, "%s %s %s", prefix, azType[pKey->aKeyAffinity[i] - SQLITE_AFF_BLOB], collation);
+    }else {
+      size = snprintf(pBuf, nBufSize, ",%s%d %s %s", prefix, i, azType[pKey->aKeyAffinity[i] - SQLITE_AFF_BLOB], collation);
+    }
+    if( size < 0 ){
+      return -1;
+    }
+    pBuf += size;
+    nBufSize -= size;
+  }
+  if( nBufSize <= 0 ){
+    return -1;
+  }
+  return 0;
+}
+
+int vectorIdxKeyNamesRender(int nKeyColumns, const char *prefix, char *pBuf, int nBufSize) {
+  int i, size;
+  for(i = 0; i < nKeyColumns && nBufSize > 0; i++){
+    if( i == 0 ){
+      size = snprintf(pBuf, nBufSize, "%s", prefix);
+    }else {
+      size = snprintf(pBuf, nBufSize, ",%s%d", prefix, i);
+    }
+    if( size < 0 ){
+      return -1;
+    }
+    pBuf += size;
+    nBufSize -= size;
+  }
+  if( nBufSize <= 0 ){
```

Comment:
isn't it possible to finish the loop above with nBufSize == 0 but still correctly?
Wouldn't a check `if( nBufSize < 0 || ( i < nKeyColumns && nBufSize == 0 ) )` be better here?

Reply:
I think we must have one free byte in the buffer because sprintf "returns the number of characters that would have been written if n had been sufficiently large, **not counting the terminating null character**."

Reply:
Makes sense. Thanks!